### PR TITLE
[Android] Add option to edit the device name

### LINKED ISSUE #2721: Custom host name setting for android devices
It would be great if there was a setting in the android client to set the device's host name. (Like there was one in the old unofficial NativeBOINC app.)

My devices got names like android_3002ca71 or android_183c46bd. Setting device name on the smartphone affects only Bluetooth and WiFi host names but not the string that BOINC client uses to identify the host.

What is even worse that the android_XXXXXXXX-type name changes after certain OS updates were received or a new ROM was flashed on the device.

It may seem to be a nice to have or purely cosmetic feature only but if you have multiple androids set up for BOINC, it could be hard to tell them apart by those ID's. A custom name would be extremely useful for that case.

PS: I tried to check for similar requests and found none. Please excuse me, if this request was discussed earlier.

### PR DESCRIPTION
**Pull request by adamradocz**
Fixes #2721

**Description of the Change**
Add option to edit the device name in Android.

**Release Notes**
[Android] Add ability to change device name.

### CODE REVIEW

**Comment by adamradocz** [comment]
`edu/berkeley/boinc/PrefsFragment.java`

Does anyone know why is this not workink? It doesn't send/write the new value.
Line 432:
```
EditText edit = dialog.findViewById(R.id.Input);
				   ClientInterfaceImplementation clientInterface = new ClientInterfaceImplementation(); //provides functions for interaction with client via rpc
				   clientInterface.setDomainName(edit.getText().toString());
				   clientInterface.close();
```

The `setDomainName` implementation is here: `edu/berkeley/boinc/rpc/RpcClient.java` at line 683.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java [hunk 1]
```diff
@@ -385,6 +388,7 @@ private void setupDialogButtons(final PrefsListItemWrapper item, final Dialog di
 		confirm.setOnClickListener(new OnClickListener() {
 			@Override
 			public void onClick(View v) {
+				// Sliders
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java]: I recommend checking your editor's indentation settings. I presume errors like this stem from a mix of `TAB`s vs. `SPACE`s and/or different `TAB`-size settings. Ideally your editor would simply auto-detect the current source file's indentation mode and use that. Atom, for instance, covers that via this [add-on](https://atom.io/packages/auto-detect-indentation).

**Comment by brevilo** [comment]
> Does anyone know why is this not working? It doesn't send/write the new value.

As far as I can see you don't open the connection at all. E.g.: `clientInterface.open(clientSocketAddress)`

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/adapter/PrefsListAdapter.java [hunk 2]
```diff
@@ -74,7 +75,9 @@ public View getView(int position, View convertView, ViewGroup parent) {
 	        	wrapper.setOnClickListener(listener);
 	    		TextView header = v.findViewById(R.id.checkbox_text);
 	    		header.setText(((PrefsListItemWrapperBool) listItem).header);
-	    	} else if(listItem instanceof PrefsListItemWrapperValue) {
+	    	}
+	    	// Numeber based items
```
[the remark this answers is not in the document]
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/adapter/PrefsListAdapter.java]: Fixed.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/adapter/PrefsListAdapter.java [hunk 3]
```diff
@@ -116,7 +119,25 @@ public View getView(int position, View convertView, ViewGroup parent) {
...
+				TextView header = v.findViewById(R.id.header);
+				header.setText(item.header);
+				TextView description = v.findViewById(R.id.description);
+				description.setText(item.description);
+
+				LinearLayout statusWrapper = v.findViewById(R.id.status_wrapper);
+				statusWrapper.setVisibility(View.VISIBLE);
+				final String value = item.status;
+				((TextView)v.findViewById(R.id.status)).setText(value);
+			}
```
[the remark this answers is not in the document]
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/adapter/PrefsListAdapter.java]: Fixed.

**Comment by adamradocz** [comment]
Added the bind service to create the Monitor object, but still doesn't send the new value to the client. I'm officially out of ideas. :(

**Comment by adamradocz** [comment]
Ok, I further investigated. It seems to me, that in the `Monitor.java` file, line 563, original 				`clientInterface.setHostInfo(model, version);` code doesn't work either. The `moder` has never written into the file, at least when I added custom value and read back, id didn't change.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java [hunk 4]
```diff
@@ -385,6 +434,7 @@ private void setupDialogButtons(final PrefsListItemWrapper item, final Dialog di
...
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java]: Unfortunately this whole block is still incorrectly indented. The `//Sliders` comment is the only correct one. All the `if` constructs that follow are aligned to `setupDialogButtons()`, not to `onClick()`. Can you please have another look? Thanks.

**Comment by brevilo** [comment]
> The `model` has never written into the file, at least when I added custom value and read back, it didn't change.

Awesome, what a royal PITA 🤦‍♂️

**Comment by brevilo** [comment]
> the original `clientInterface.setHostInfo(model, version);` code doesn't work either

Ok, let's tackle this one by one:
* I can't see anything obviously wrong with your code (`setHostInfo()`
* How about the logs? Do you see your request in there? `sendRequest()` logs all of them
* If so, it's I'd say the problem is at the receiving end of the RPC.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java (hunk 4 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java]: First let's have a working version, then I'll work on the indentation later.

**Comment by JuhaSointusalo** [comment]
[Hint](https://github.com/BOINC/boinc/blob/master/client/gui_rpc_server_ops.cpp#L382)

**Comment by adamradocz** [comment]
> [Hint](https://github.com/BOINC/boinc/blob/master/client/gui_rpc_server_ops.cpp#L382)

Thanks, that helped a lot! :)

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java (hunk 4 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java]: Fixed.

**Comment by adamradocz** [comment]
It's working. Yay! :)
The only issue now, that the preference item not refreshed when the new value is set, so the new value is only visible if you navigate to another menu and back.

**Comment by adamradocz** [comment]
Done. Please review it ASAP, because this section needs refactoring and I'll make another PR for that.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java [hunk 5]
```diff
@@ -675,6 +675,35 @@ public synchronized boolean setHostInfo(String hostInfo, String version){
...
+		try {
+			sendRequest(mRequest.toString());
+			SimpleReplyParser parser = SimpleReplyParser.parse(receiveReply());
+			if (parser == null)
+				return false;
+			mLastErrorMessage = parser.getErrorMessage();
+			return parser.result();
+		}
+		catch (IOException e) {
+			if(Logging.WARNING) Log.w(Logging.TAG, "error in networkAvailable()", e);
```
[the remark this answers is not in the document]
Comment (brevilo) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java]: It has the same copied log message referring to the wrong method.

[the remark this answers is not in the document]
Comment (brevilo) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java]: * https://github.com/BOINC/boinc/pull/2908/files/643e12855592afc7197fceb502dc2dd3ab04d3c8#diff-755e90dbd3ff372b05683a1102aa379dR641
* https://github.com/BOINC/boinc/pull/2908/files/643e12855592afc7197fceb502dc2dd3ab04d3c8#diff-755e90dbd3ff372b05683a1102aa379dR673
* https://github.com/BOINC/boinc/pull/2908/files/643e12855592afc7197fceb502dc2dd3ab04d3c8#diff-755e90dbd3ff372b05683a1102aa379dR702

[the remark this answers is not in the document]
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java]: Done.

**Comment by brevilo** [comment]
> If so, I'd say the problem is at the receiving end of the RPC.
Apparently it was 😁

**Pull request merged**